### ISSUE TITLE
6300 coupon codes in Commerce results in alot of out-of-memory issues.

### ISSUE THREAD

**Issue by ragnarfrosti** [craftcms/commerce#2164]
**Description**
One of our clients use coupon codes a lot. They added about 6300 coupon codes to Commerce couple of days ago and that has resulted in a huge problem on their website as the customers are receiving a lot of Out of memory issues since then.

The website has 512MB memory limit at the moment and 4GB memory in total.

We have tried to understand how the amount of coupon codes can affect the whole webshop so much and the problem seems to be that Commerce always loads all active coupon codes in the system to a variable before checking if a coupon code exist, in adjustments and basically whenever there is any coupon code related checks.

With 6300 coupon code objects in the memory, looping through these takes to much resources from the system.

It's not possible at the moment to add a coupon code on the Cart as the function getDiscountByCode in the updateCart function loads all the coupon codes to the memory before checking if the code exist....

```
public function getDiscountByCode($code)
{
    if (!$code) {
        return null;
    }

    return ArrayHelper::firstWhere($this->getAllDiscounts(), function($discount) use ($code) {
        return ($discount->code && $code && (strcasecmp($code, $discount->code) == 0));
    });
}
```
(In our client case, where many of the discount codes have over 25 purchasables applied to each code, the getAllDiscounts() function returns almost 100000 results from the database which then populate the 6300 discount objects.)

And even when saving the cart without having any coupon code applied on the cart the populateFromPurchaseable function in a lineItem fetches and loop through all the Discount codes to check if it's valid on the order...

```
public function populateFromPurchasable(PurchasableInterface $purchasable)
{
    $this->price = $purchasable->getPrice();
    $this->salePrice = Plugin::getInstance()->getSales()->getSalePriceForPurchasable($purchasable, $this->order);
    $this->taxCategoryId = $purchasable->getTaxCategoryId();
    $this->shippingCategoryId = $purchasable->getShippingCategoryId();
    $this->sku = $purchasable->getSku();
    $this->description = $purchasable->getDescription();

    // Check to see if there is a discount applied that ignores Sales for this line item
    $ignoreSales = false;
    foreach (Plugin::getInstance()->getDiscounts()->getAllActiveDiscounts($this->getOrder()) as $discount) {
        if ($discount->enabled && Plugin::getInstance()->getDiscounts()->matchLineItem($this, $discount, true)) {
            $ignoreSales = $discount->ignoreSales;
            if ($ignoreSales) {
                break;
            }
        }
    }
    ......
}
```

And the same applies to the adjusters, when calculating the adjusters on a order, all coupon codes are fetched and looped through for the shipping adjustment and maybe somewhere else.

### To summarize
All the discounts in the system seems to unnecessarily be fetched and saved to a variable for comparisons which requires a huge amount of memory from the server when there are 6300 coupon codes in the system.

PS. The Commerce > Promotions > Discounts loads all the coupon codes in one query as well so that one doesn't work that well either ;)

**Additional info**
- Craft version: Craft CMS 3.5.14
- Commerce version: 3.2.11
- PHP version: 7.3.27
- Database driver & version: MySQL 5.5.5

**Comment by lukeholder** [comment]
We can look to make some performance improvements, but out of interest, how were that many discounts generated? By hand? Programmatically? 

We plan split coupons out from discounts as mentioned here: #1388 if we did that, would you still need that many discounts?

Looking into the performance issues with `getDiscountByCode`  and `getAllActiveDiscounts()` and will see if we can make some immediate improvements.

**Comment by ragnarfrosti** [comment]
Thanks for looking into this @lukeholder.

Our client in this case has used the coupon codes quite a lot the last couple of years and created personalized coupon codes for it’s customers for different events, so we built a custom plugin for them so they can export the list of all the discounts or upload a CSV-file to bulk import/delete the discounts.

> We plan split coupons out from discounts as mentioned here: #1388 if we did that, would you still need that many discounts?

Depending on how this would work it might solve some of the issues in our case. In most cases they want unique codes for each person, so one discount with 4000 unique codes would work in that case.

> Looking into the performance issues with getDiscountByCode and getAllActiveDiscounts() and will see if we can make some immediate improvements.

It would be great if this could be improved in some way as our client had/has plans to import over 25000 codes the next couple of weeks... :see_no_evil:
